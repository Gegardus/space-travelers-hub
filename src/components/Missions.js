import React from 'react';

const Missions = () => (
  <main>
    <h1>Hello from missions</h1>
  </main>
);

export default Missions;
