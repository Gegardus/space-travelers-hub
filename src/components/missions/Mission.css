.mission-container {
  padding: 2rem;
}

.main-title,
.main-title ul li {
  background-color: #fff;
  border: solid 1px rgba(0, 0, 0, 0.096);
}

.main-title ul li {
  padding: 5px;
}

.mission-title ul,
.main-title ul {
  display: grid;
  grid-template-columns: 1fr 3fr 1fr 1fr;
  list-style: none;
  margin-bottom: 0;
  padding-left: 0;
}

.mission-title ul:nth-of-type(2n + 1) {
  background-color: #f2f2f2;
}

.mission-title ul li {
  border: solid 1px rgba(0, 0, 0, 0.096);
  padding: 5px;
}

.title {
  font-weight: bold;
  font-size: 1.3rem;
}

.name {
  font-weight: bold;
}

.status {
  display: flex;
  align-items: center;
  justify-content: center;
}

.loading {
  display: flex;
  align-items: center;
  justify-content: center;
}

.not-active {
  background-color: #6d757d;
  color: #fff;
  padding: 2px 5px;
  border-radius: 5px;
  font-weight: bold;
}

.active-member {
  background-color: #17a2b9;
  color: #fff;
  padding: 2px 5px;
  border-radius: 5px;
  font-weight: bold;
}
