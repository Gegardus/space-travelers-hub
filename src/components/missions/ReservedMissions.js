import React from 'react';
import { useSelector } from 'react-redux';

const ReservedMissions = () => {
  const selectedMissions = useSelector((state) => state.missions);
  const reservedMissions = selectedMissions.filter((mission) => mission.reserved === true);
  return (
    <ul className="profile-container">
      <h4>My Missions</h4>
      {
 reservedMissions.map((mission) => (
   <li key={mission.mission_id}>
     <p>{mission.mission_name}</p>
   </li>
 ))
}
    </ul>
  );
};

export default ReservedMissions;
