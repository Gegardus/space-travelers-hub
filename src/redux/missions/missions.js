const ADD_MISSION = 'spacetravelers-hub/redux/missions/ADD_MISSION';
const missionApiURL = 'https://api.spacexdata.com/v3/missions';

const initialState = [];

export const addMission = (missions) => ({
  type: ADD_MISSION,
  missions,
});

export const getMissionFromAPI = () => (async (dispatch) => {
  const response = await fetch(missionApiURL);
  const data = await response.json();
  const missionsArr = [];
  data.forEach((d) => {
    const mission = {
      id: d.mission_id,
      name: d.mission_name,
      description: d.description,
      join: false,
    };
    missionsArr.push(mission);
  });
  dispatch(addMission(missionsArr));
});

const missionReducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD_MISSION:
      return [...state, ...action.missions];
    default:
      return state;
  }
};

export default missionReducer;
