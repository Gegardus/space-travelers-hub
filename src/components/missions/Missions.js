import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Button from 'react-bootstrap/Button';
import {
  getMissionFromAPI,
  reserveMission,
  cancelMissionReservation,
} from '../../redux/missions/missions';
import './Mission.css';

const Missions = () => {
  const dispatch = useDispatch();
  const missionsData = useSelector((state) => state.missions);
  const getMissions = () => {
    if (missionsData.length === 0) {
      dispatch(getMissionFromAPI());
    }
  };
  useEffect(() => {
    getMissions();
  }, []);

  return (
    <div className="mission-container">
      <div className="main-title">
        <ul>
          <li className="name title">Mission</li>
          <li className="title">Description</li>
          <li className="status title">Status</li>
          <li> </li>
        </ul>
      </div>
      <div className="mission-title">
        {missionsData.map((mission) => (
          <ul key={mission.mission_id}>
            <li className="name title">{mission.mission_name}</li>
            <li className="description">{mission.description}</li>
            <li className="status">
              {mission.reserved ? <div className="active-member">Active Member</div> : <div className="not-active">Not A Member</div>}
            </li>
            <li className="loading">
              {mission.reserved ? (
                <Button
                  variant="outline-danger"
                  onClick={() => dispatch(cancelMissionReservation(mission.mission_id))}
                >
                  Leave Mission
                </Button>
              ) : (
                <Button
                  variant="outline-secondary"
                  onClick={() => dispatch(reserveMission(mission.mission_id))}
                >
                  Join Mission
                </Button>
              )}
            </li>
          </ul>
        ))}
      </div>
    </div>
  );
};

export default Missions;
