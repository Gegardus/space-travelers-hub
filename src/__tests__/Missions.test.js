import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import store from '../redux/configureStore';
import Missions from '../components/missions/Missions';
import { addMission, reserveMission, cancelMissionReservation } from '../redux/missions/missions';

describe('Mission Component test', () => {
  it('Renders Missions Page', () => {
    const app = renderer.create(
      <Provider store={store}>
        <Missions />
      </Provider>,
    ).toJSON();
    expect(app).toMatchSnapshot();
  });

  it('has a text in the DOM', () => {
    const { container } = render(<Provider store={store}><Missions /></Provider>);
    expect(container.getElementsByClassName('mission-container').length).toBe(1);
  });
  it('Returns an addMission() action object ', () => {
    expect(addMission([])).toStrictEqual({ missions: [], type: 'spacetravelers-hub/redux/missions/ADD_MISSION' });
  });
  it('Returns a reserveMission() action object ', () => {
    expect(reserveMission('F4F83DE')).toEqual({ type: 'spacetravelers-hub/redux/missions/RESERVE_MISSION', missionId: 'F4F83DE' });
  });
  it('Returns a cancelMissionReservation() action object ', () => {
    expect(cancelMissionReservation('F4F83DE')).toEqual({ type: 'spacetravelers-hub/redux/missions/CANCEL_RESERVATION', missionId: 'F4F83DE' });
  });
});
