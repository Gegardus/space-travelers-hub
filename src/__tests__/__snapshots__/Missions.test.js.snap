// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission Component test Renders Missions Page 1`] = `
<div
  className="mission-container"
>
  <div
    className="main-title"
  >
    <ul>
      <li
        className="name title"
      >
        Mission
      </li>
      <li
        className="title"
      >
        Description
      </li>
      <li
        className="status title"
      >
        Status
      </li>
      <li>
         
      </li>
    </ul>
  </div>
  <div
    className="mission-title"
  />
</div>
`;
